pymmcore_plus/core/_property.py

```diff
@@ -0,0 +1,169 @@
+        return self._mmc
+
+    @property
+    def value(self) -> Any:
+        """Return current property value, cast to appropriate type if applicable."""
+        v = self._mmc.getProperty(self.device, self.name)
+        if type_ := self.type().to_python():
+            v = type_(v)
+        return v
+
+    @value.setter
+    def value(self, val: Any) -> None:
+        """Set current property value"""
+        self.setValue(val)
+
+    def setValue(self, val: Any) -> None:
+        """Functional alternate to property setter."""
+        if self.isReadOnly():
+            import warnings
+
+            warnings.warn(f"'{self.device}::{self.name}' is a read-only property.")
+        self._mmc.setProperty(self.device, self.name, val)
+
+    def isReadOnly(self) -> bool:
+        """Return `True` if property is read only."""
+        return self._mmc.isPropertyReadOnly(self.device, self.name)
+
+    def isPreInit(self) -> bool:
+        """Return `True` if property must be defined prior to initialization."""
+        return self._mmc.isPropertyPreInit(self.device, self.name)
+
+    def hasLimits(self) -> bool:
+        """Return `True` if property has limits"""
+        return self._mmc.hasPropertyLimits(self.device, self.name)
+
+    def lowerLimit(self) -> float:
+        """Return lower limit if property has limits, or 0 otherwise."""
+        return self._mmc.getPropertyLowerLimit(self.device, self.name)
+
+    def upperLimit(self) -> float:
+        """Return upper limit if property has limits, or 0 otherwise."""
+        return self._mmc.getPropertyUpperLimit(self.device, self.name)
+
+    def range(self) -> Tuple[float, float]:
+        """Return (lowerLimit, upperLimit) range tuple."""
+        return (self.lowerLimit(), self.upperLimit())
+
+    def type(self) -> PropertyType:
+        """Return `PropertyType` of this property."""
+        return self._mmc.getPropertyType(self.device, self.name)
+
+    def deviceType(self) -> DeviceType:
+        """Return `DeviceType` of the device owning this property."""
+        return self._mmc.getDeviceType(self.device)
+
+    def allowedValues(self) -> Tuple[str, ...]:
+        """Return allowed values for this property, if contstrained."""
+        # https://github.com/micro-manager/mmCoreAndDevices/issues/172
+        allowed = self._mmc.getAllowedPropertyValues(self.device, self.name)
+        if not allowed and self.deviceType() is DeviceType.StateDevice:
+            if self.name == g_Keyword_State:
+                n_states = self._mmc.getNumberOfStates(self.device)
+                allowed = tuple(str(i) for i in range(n_states))
+            elif self.name == g_Keyword_Label:
+                allowed = self._mmc.getStateLabels(self.device)
+        return allowed
+
+    def dict(self) -> InfoDict:
+        """Return dict of info about this Property.
+
+        Contains the following keys (See `InfoDict` type): "valid", "value", "type",
```

Comment:
Is there any way for a user to access the `InfoDict` type, especially given being hidden insid ethe type_checking if?

Reply:
Not as an import outside of a Type checking block... (but they could import inside of a type_checking block too).  what sort of usage are you picturing?

Reply:
moved outside the type_checking clause just in case